### PR TITLE
[SYCLomatic] Support migration of 4 cub::{StoreDirectBlocked, StoreDirectStriped} and cub::{BlockLoad, BlockStore}::* APIs

### PR DESCRIPTION
**Pull request by yihwang-nv**

### CODE REVIEW

File: clang/lib/DPCT/CUBAPIMigration.cpp [hunk 1]
```diff
@@ -252,9 +254,17 @@ void CubMemberCallRule::runRule(
         Name == "BlockedToStriped" || Name == "StripedToBlocked" ||
         Name == "StripedToBlocked" || Name == "ScatterToBlocked" ||
         Name == "ScatterToStriped";
-    if (isBlockRadixSort || isBlockExchange) {
-      std::string HelpFuncName =
-          isBlockRadixSort ? "group_radix_sort" : "exchange";
+    if (isBlockRadixSort || isBlockExchange || Name == "Load" ||
+        Name == "Store") {
+      std::string HelpFuncName;
+      if (isBlockRadixSort)
```
Comment (tangjj11) [inline clang/lib/DPCT/CUBAPIMigration.cpp:260]: may use a unordered map to do this?

File: clang/lib/DPCT/Rewriters/CUB/RewriterUtilityFunctions.cpp [hunk 2]
```diff
@@ -165,14 +165,33 @@ RewriterMap dpct::createUtilityFunctionsRewriterMap() {
...
           HeaderType::HT_DPCT_GROUP_Utils,
           CALL_FACTORY_ENTRY(
               "cub::LoadDirectStriped",
               CALL(PRETTY_TEMPLATED_CALLEE(MapNames::getDpctNamespace() +
                                                "group::load_direct_striped",
-                                           0, 1, 2, 3),
-                   ARG(0), ARG(1), ARG(2))))};
+                                           1, 2, 3),
+                   NDITEM, ARG(1), ARG(2))))
+
```
Comment (tangjj11) [inline clang/lib/DPCT/Rewriters/CUB/RewriterUtilityFunctions.cpp:178]: need not empty line?

File: clang/runtime/dpct-rt/include/dpct/group_utils.hpp [hunk 3]
```diff
@@ -750,54 +750,92 @@ __dpct_inline__ void load_striped(const Item &item, InputIteratorT block_itr,
...
-template <typename InputT, size_t ElementsPerWorkItem, typename InputIteratorT>
-__dpct_inline__ void load_direct_blocked(size_t linear_tid,
-                                         InputIteratorT block_itr,
-                                         InputT (&items)[ElementsPerWorkItem]) {
+/// \param item The calling work-item.
+/// \param input_iter The work-group's base input iterator for loading from.
+/// \param data Data to load.
+template <typename T, size_t ElementsPerWorkItem, typename InputIteratorT,
+          typename ItemT>
+__dpct_inline__ void load_direct_blocked(const ItemT &item,
```
Comment (tangjj11) [inline clang/runtime/dpct-rt/include/dpct/group_utils.hpp:765]: can we change the API signature in header file?

File: clang/runtime/dpct-rt/include/dpct/group_utils.hpp [hunk 4]
```diff
@@ -750,54 +750,92 @@ __dpct_inline__ void load_striped(const Item &item, InputIteratorT block_itr,
...
-          typename InputIteratorT>
-__dpct_inline__ void load_direct_striped(size_t linear_tid,
-                                         InputIteratorT block_itr,
-                                         InputT (&items)[ElementsPerWorkItem]) {
+/// \tparam T The data type to store.
+/// \tparam ElementsPerWorkItem The number of consecutive elements partitioned
+/// onto each work-item.
+/// \tparam OutputIteratorT  The random-access iterator type for output.
+/// \iterator.
+/// \param linear_id A suitable linear identifier for the calling work-item.
```
Comment (tangjj11) [inline clang/runtime/dpct-rt/include/dpct/group_utils.hpp]: wrong comment

**Review by tangjj11 — CHANGES_REQUESTED** [review]

File: clang/runtime/dpct-rt/include/dpct/group_utils.hpp (hunk 3 above)
Comment (yihwang-nv) [reply clang/runtime/dpct-rt/include/dpct/group_utils.hpp:765]: The API signature not changed

File: clang/lib/DPCT/CUBAPIMigration.cpp (hunk 1 above)
Comment (yihwang-nv) [reply clang/lib/DPCT/CUBAPIMigration.cpp:260]: I don't think so. inline string has more optimization opportunities.

**Review by tangjj11 — APPROVED** [review]

File: clang/runtime/dpct-rt/include/dpct/group_utils.hpp [hunk 5]
```diff
@@ -750,54 +750,92 @@ __dpct_inline__ void load_striped(const Item &item, InputIteratorT block_itr,
...
+/// \tparam T The data type to store.
+/// \tparam ElementsPerWorkItem The number of consecutive elements partitioned
+/// onto each work-item.
+/// \tparam OutputIteratorT  The random-access iterator type for output.
+/// \iterator.
+/// \param item The calling work-item.
+/// \param output_iter The work-group's base output iterator for writing.
+/// \param data Data to store.
+template <typename T, size_t ElementsPerWorkItem, typename OutputIteratorT,
+          typename ItemT>
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/group_utils.hpp:810]: no comment for ItemT?

File: clang/runtime/dpct-rt/include/dpct/group_utils.hpp [hunk 6]
```diff
@@ -750,54 +750,92 @@ __dpct_inline__ void load_striped(const Item &item, InputIteratorT block_itr,
...
+/// segment of items.
 ///
-/// \param linear_tid A suitable linear identifier for the calling work-item.
-///
-/// \param block_itr The work-group's base input iterator for loading from.
+/// \tparam T The data type to store.
+/// \tparam ElementsPerWorkItem The number of consecutive elements partitioned
+/// onto each work-item.
+/// \tparam OutputIteratorT  The random-access iterator type for output.
+/// \iterator.
```
Comment (abhilash1910) [inline clang/runtime/dpct-rt/include/dpct/group_utils.hpp:804]:
```suggestion
/// \iterator.
/// \tparam ItemT The type of the sycl work item 
```

File: clang/runtime/dpct-rt/include/dpct/group_utils.hpp [hunk 7]
```diff
@@ -750,54 +750,92 @@ __dpct_inline__ void load_striped(const Item &item, InputIteratorT block_itr,
...
-template <size_t WorkGroupSize, typename InputT, int ElementsPerWorkItem,
-          typename InputIteratorT>
-__dpct_inline__ void load_direct_striped(size_t linear_tid,
-                                         InputIteratorT block_itr,
-                                         InputT (&items)[ElementsPerWorkItem]) {
+/// \tparam T The data type to store.
+/// \tparam ElementsPerWorkItem The number of consecutive elements partitioned
+/// onto each work-item.
+/// \tparam OutputIteratorT  The random-access iterator type for output.
+/// \iterator.
```
Comment (abhilash1910) [inline clang/runtime/dpct-rt/include/dpct/group_utils.hpp:829]:
```suggestion
/// \iterator.
/// \tparam ItemT The type of the sycl work item 
```

File: clang/runtime/dpct-rt/include/dpct/group_utils.hpp [hunk 8]
```diff
@@ -750,54 +753,100 @@ __dpct_inline__ void load_striped(const Item &item, InputIteratorT block_itr,
...
-///
-/// \param block_itr The work-group's base input iterator for loading from.
-///
-/// \param items Data to load
-template <typename InputT, size_t ElementsPerWorkItem, typename InputIteratorT>
-__dpct_inline__ void load_direct_blocked(size_t linear_tid,
-                                         InputIteratorT block_itr,
-                                         InputT (&items)[ElementsPerWorkItem]) {
+/// \tparam ItemT The execution instance’s point in the iteration space, usually
+/// sycl::id, sycl::item, sycl::nd_item, etc...
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/group_utils.hpp]: sycl::nd_item,  index space class

**Review by zhimingwang36 — APPROVED** [review]

**Pull request merged**